### PR TITLE
SDI-623 Function to map updated course rules

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleServiceTest.kt [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.SlotCategory
+import java.time.LocalDate
+import java.time.LocalTime
+
+class ScheduleRuleServiceTest {
+
+  private val scheduleRuleService = ScheduleRuleService()
+
+  @Nested
+  inner class CreateScheduleRules {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleServiceTest.kt:20]: Corrected the name of the top level class but GitHub is rubbish and thinks it's brand new - but CreateScheduleRules has not been changed at all.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleServiceTest.kt [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        }
+        with(newScheduleRules[1]) {
+          assertThat(startTime).isEqualTo("2022-10-01T08:00")
+        }
+      }
+    }
+  }
+
+  @Nested
+  inner class BuildNewRules {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleServiceTest.kt:112]: The new tests are all in this class

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt [hunk 3]
```diff
@@ -38,4 +39,40 @@ class ScheduleRuleService {
...
   }
+
+  fun buildNewRules(requestedRules: List<ScheduleRuleRequest>, existingActivity: CourseActivity): List<CourseScheduleRule> {
+    val monthStart = LocalDate.now().with(TemporalAdjusters.firstDayOfMonth())
+    return requestedRules.map {
+      validateRequest(it)
+      CourseScheduleRule(
+        id = findExistingRuleId(it, existingActivity.courseScheduleRules),
+        courseActivity = existingActivity,
+        startTime = LocalDateTime.of(monthStart, it.startTime),
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt:50]: What's all this start of the month stuff?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt [hunk 4]
```diff
@@ -38,4 +39,40 @@ class ScheduleRuleService {
...
+        it.monday == requestedRule.monday &&
+        it.tuesday == requestedRule.tuesday &&
+        it.wednesday == requestedRule.wednesday &&
+        it.thursday == requestedRule.thursday &&
+        it.friday == requestedRule.friday &&
+        it.saturday == requestedRule.saturday &&
+        it.sunday == requestedRule.sunday
+    }
+      ?.id
+      ?: let { 0 }
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt:77]: I guess 0 means create new rule??

**Review by andymarke — APPROVED** [review]
Lgtm

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt (hunk 3 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt:50]: Nomis models the start and end times as ISO date time (even though it's only interested in the times themselves). Oracle shenanigans methinks.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt (hunk 4 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleRuleService.kt:77]: Yes, for Hibernate.

**Pull request merged**